### ISSUE TITLE
ActiveRecord first() and last() behave differently with limit()

### ISSUE THREAD

**Issue by brchristian** [rails/rails#23979]
### Steps to reproduce

``` ruby
5.times do
  Topic.create!
end

Topic.limit(1).first(2)
  Topic Load (0.8ms)  SELECT  "topics".* FROM "topics" ORDER BY "topics"."id" ASC LIMIT $1  [["LIMIT", 2]]
=> [#<Topic id: 1>, #<Topic id: 2>]

Topic.limit(1).last(2)
  Topic Load (0.3ms)  SELECT  "topics".* FROM "topics" LIMIT $1  [["LIMIT", 1]]
=> [#<Topic id: 5>]
```
### Expected behavior

I would expect that, for any ActiveRecord relation, `first(n).size` and `last(n).size` would always be equal.
### Actual behavior

The `first(n)` method seems to "overwrite" the `limit(1)`, whereas the `last(n)` method respects it. As a result, `Topic.limit(1).first(2).size` is 2 and `Topic.limit(1).last(2).size` is 1.

In my opinion the behavior of the `last(n)` method is correct. This is consistent with DHH's comments here, for instance: https://github.com/rails/rails/pull/23598#issuecomment-189675440.

This would seem to suggest that the `first(n)` method should ought to return an array of size 1 in the case of `Topic.limit(1).first(2)`.
### System configuration

**Rails version**: current `master` (https://github.com/rails/rails/commit/75097933e26e5147f666f485d41b38967c76311c)

**Ruby version**: 2.3.0

**Comment by mechanicles** [comment]
I also confirmed this issue. `first` and `last` methods behave differently while using it with `limit`. This happens with gem (4.2.0) as well as with master.

https://gist.github.com/mechanicles/e8192f3d24aa1c186aee

Output:

```
Finished in 0.016156s, 61.8960 runs/s, 123.7919 assertions/s.

  1) Failure:
BugTest#test_limit_with_last [limit_last_test.rb:42]:
Expected: #<Post id: 1>
  Actual: #<Post id: 3>

1 runs, 2 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by brchristian** [comment]
@mechanicles What you are seeing in your test may also be related to https://github.com/rails/rails/issues/23607 as well, where `first` imposes primary key order but `last` does not.

**Comment by mechanicles** [comment]
@brchristian This issue also got fixed in this PR #24000 too
